File: go/test/conan_package_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/gitlab-sdk/go"
	"github.com/voxgig-sdk/gitlab-sdk/go/core"

	vs "github.com/voxgig-sdk/gitlab-sdk/go/utility/struct"
)

func TestConanPackageEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ConanPackage(nil)
		if ent == nil {
			t.Fatal("expected non-nil ConanPackageEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := conan_packageBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "conan_package." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_CONAN_PACKAGE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		conanPackageRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.conan_package", setup.data)))
		var conanPackageRef01Data map[string]any
		if len(conanPackageRef01DataRaw) > 0 {
			conanPackageRef01Data = core.ToMapAny(conanPackageRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = conanPackageRef01Data

		// UPDATE
		conanPackageRef01Ent := client.ConanPackage(nil)
		conanPackageRef01DataUp0Up := map[string]any{
			"file_id": setup.idmap["file_id"],
			"package_channel": setup.idmap["package_channel"],
			"package_username": setup.idmap["package_username"],
			"package_version": setup.idmap["package_version"],
			"recipe_revision": setup.idmap["recipe_revision"],
		}

		conanPackageRef01ResdataUp0Result, err := conanPackageRef01Ent.Update(conanPackageRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		conanPackageRef01ResdataUp0 := core.ToMapAny(conanPackageRef01ResdataUp0Result)
		if conanPackageRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}

		// LOAD
		conanPackageRef01MatchDt0 := map[string]any{}
		conanPackageRef01DataDt0Loaded, err := conanPackageRef01Ent.Load(conanPackageRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if conanPackageRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

		// REMOVE
		conanPackageRef01MatchRm0 := map[string]any{
			"id": conanPackageRef01Data["id"],
		}
		_, err = conanPackageRef01Ent.Remove(conanPackageRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

	})
}

func conan_packageBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "conan_package", "ConanPackageTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read conan_package test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse conan_package test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"conan_package01", "conan_package02", "conan_package03", "project01", "project02", "project03", "conan01", "conan02", "conan03", "file01", "file02", "file03", "export01", "export02", "export03", "package01", "package02", "package03", "revision01", "revision02", "revision03", "package_channel01", "package_username01", "package_version01", "recipe_revision01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITLAB_TEST_CONAN_PACKAGE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITLAB_TEST_CONAN_PACKAGE_ENTID": idmap,
		"GITLAB_TEST_LIVE":      "FALSE",
		"GITLAB_TEST_EXPLAIN":   "FALSE",
		"GITLAB_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GITLAB_TEST_CONAN_PACKAGE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add file_id alias for update test.
	if idmapResolved["file_id"] == nil {
		idmapResolved["file_id"] = idmapResolved["file01"]
	}
	// Add package_channel alias for update test.
	if idmapResolved["package_channel"] == nil {
		idmapResolved["package_channel"] = idmapResolved["package_channel01"]
	}
	// Add package_username alias for update test.
	if idmapResolved["package_username"] == nil {
		idmapResolved["package_username"] = idmapResolved["package_username01"]
	}
	// Add package_version alias for update test.
	if idmapResolved["package_version"] == nil {
		idmapResolved["package_version"] = idmapResolved["package_version01"]
	}
	// Add recipe_revision alias for update test.
	if idmapResolved["recipe_revision"] == nil {
		idmapResolved["recipe_revision"] = idmapResolved["recipe_revision01"]
	}

	if env["GITLAB_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GITLAB_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGitlabSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITLAB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITLAB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/ConanPackageEntityTest.php
<?php
declare(strict_types=1);

// ConanPackage entity test

require_once __DIR__ . '/../gitlab_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class ConanPackageEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = GitlabSDK::test(null, null);
        $ent = $testsdk->ConanPackage(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = conan_package_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["update", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "conan_package." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_CONAN_PACKAGE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $conan_package_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.conan_package")));
        $conan_package_ref01_data = null;
        if (count($conan_package_ref01_data_raw) > 0) {
            $conan_package_ref01_data = Helpers::to_map($conan_package_ref01_data_raw[0][1]);
        }

        // UPDATE
        $conan_package_ref01_ent = $client->ConanPackage(null);
        $conan_package_ref01_data_up0_up = [
            "file_id" => $setup["idmap"]["file_id"],
            "package_channel" => $setup["idmap"]["package_channel"],
            "package_username" => $setup["idmap"]["package_username"],
            "package_version" => $setup["idmap"]["package_version"],
            "recipe_revision" => $setup["idmap"]["recipe_revision"],
        ];

        $conan_package_ref01_resdata_up0_result = $conan_package_ref01_ent->update($conan_package_ref01_data_up0_up, null);
        $conan_package_ref01_resdata_up0 = Helpers::to_map($conan_package_ref01_resdata_up0_result);
        $this->assertNotNull($conan_package_ref01_resdata_up0);

        // LOAD
        $conan_package_ref01_match_dt0 = [];
        $conan_package_ref01_data_dt0_loaded = $conan_package_ref01_ent->load($conan_package_ref01_match_dt0, null);
        $this->assertNotNull($conan_package_ref01_data_dt0_loaded);

    }
}

function conan_package_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/conan_package/ConanPackageTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = GitlabSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["conan_package01", "conan_package02", "conan_package03", "project01", "project02", "project03", "conan01", "conan02", "conan03", "file01", "file02", "file03", "export01", "export02", "export03", "package01", "package02", "package03", "revision01", "revision02", "revision03", "package_channel01", "package_username01", "package_version01", "recipe_revision01"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("GITLAB_TEST_CONAN_PACKAGE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "GITLAB_TEST_CONAN_PACKAGE_ENTID" => $idmap,
        "GITLAB_TEST_LIVE" => "FALSE",
        "GITLAB_TEST_EXPLAIN" => "FALSE",
        "GITLAB_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["GITLAB_TEST_CONAN_PACKAGE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }
    if (!isset($idmap_resolved["file_id"])) {
        $idmap_resolved["file_id"] = $idmap_resolved["file01"];
    }
    if (!isset($idmap_resolved["package_channel"])) {
        $idmap_resolved["package_channel"] = $idmap_resolved["package_channel01"];
    }
    if (!isset($idmap_resolved["package_username"])) {
        $idmap_resolved["package_username"] = $idmap_resolved["package_username01"];
    }
    if (!isset($idmap_resolved["package_version"])) {
        $idmap_resolved["package_version"] = $idmap_resolved["package_version01"];
    }
    if (!isset($idmap_resolved["recipe_revision"])) {
        $idmap_resolved["recipe_revision"] = $idmap_resolved["recipe_revision01"];
    }

    if ($env["GITLAB_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["GITLAB_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new GitlabSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["GITLAB_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["GITLAB_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
